### PR TITLE
[Form Provider Refactor] IdologyQuestions

### PR DESCRIPTION
**Pull request by kowczarz**
### Fixed Issues

$ https://github.com/Expensify/App/issues/30313
$ https://github.com/Expensify/App/issues/31612

### Tests
1. Go to staging.new Expensify.com or open Expensify App
2. Login with expensifail account
3. Go to Settings > Wallet > Enable Wallet
4. Go through plaid verification flow
5. Follow the flow until Additional details
6. Fill the form with following data: ```First name: Alberta
Last name: Bobbeth
SSN: 3333
Any other information for the rest of the fields```
7. Press Save & continue
8. Idology questions should appear
9. Hit back button
10. Additional details should be filled with draft values
11. Press Save & continue
12. Don't fill anything and press Save & continue
13. Errors should appear
14. Select any response
15. Error should disappear
16. You should be able to proceed to the next question

- [x] Verify that no errors appear in the JS console

### Offline tests

Same as above

### QA Steps

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
        - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/91068263/d1e70ddf-25fa-451d-a330-803690c0b0b3

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/91068263/62d3da08-c5ff-4799-93cf-50e57e224ef4

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/91068263/2a7959e9-8bde-4868-b66d-0302eab5c1c8

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/91068263/c05e42b5-e4d5-447c-b604-1dc47840d468

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/91068263/816aff3d-f88f-4ac4-8ec6-776deb448472

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/91068263/f4728126-a6af-4571-9c88-32febb63ed41

</details>

### CODE REVIEW

**Review by blazejkustra — APPROVED** [review]
LGTM!

File: src/pages/EnablePayments/IdologyQuestions.js
Comment (s77rt) [inline src/pages/EnablePayments/IdologyQuestions.js:139]:
```suggestion
```
`isLoading` is not a prop

File: src/pages/EnablePayments/IdologyQuestions.js
Comment (s77rt) [inline src/pages/EnablePayments/IdologyQuestions.js:140]:
```suggestion
```

This is causing a double offline indicator (not needed)

**Review by s77rt — CHANGES_REQUESTED** [review]

File: src/pages/EnablePayments/IdologyQuestions.js
Comment (s77rt) [inline src/pages/EnablePayments/IdologyQuestions.js:133]: 1. Remove `submitButtonStyles` prop
2. Change style to `{[styles.flexGrow1, styles.ph5]}`
3. Removing the wrapping view inside

It should look something like the following (assuming the rest of the comments are solved):

```js
            <FormProvider
                formID="walletAdditionalDetails"
                validate={validate}
                onSubmit={submitAnswers}
                scrollContextEnabled
                style={[styles.flexGrow1, styles.ph5]}
                submitButtonText={translate('common.saveAndContinue')}
            >
                <InputWrapper
                    InputComponent={SingleChoiceQuestion}
                    inputID="answer"
                    prompt={currentQuestion.prompt}
                    possibleAnswers={possibleAnswers}
                    currentQuestionIndex={currentQuestionIndex}
                    onValueChange={chooseAnswer}
                />
            </FormProvider>

```

File: src/components/Form/FormProvider.js
Comment (kowczarz) [reply src/components/Form/FormProvider.js:52]: Yes, `IdologyQuestions` doesn't use drafts but during testing the wallet flow I spotted that it's missing and I thought It can be fixed here.

Comment (s77rt) [reply src/components/Form/FormProvider.js:52]: Ah I see. Indeed the FormProvider is missing the use of drafts values but can you please specify a test scenario to make sure this is working as intended? (update PR description testing steps - test steps should cover all the changes in the PR)

File: src/components/Form/FormProvider.js
Comment (s77rt) [inline src/components/Form/FormProvider.js:219]: What is the use for the last condition `inputValues[inputID] === undefined`? Is there a case where `inputValues[inputID]` would have a value when calling `registerInput`?

File: src/components/Form/FormProvider.js
Comment (kowczarz) [reply src/components/Form/FormProvider.js:77]: But without `customErrorMessage` it's not possible to keep current layout of the form.
<details>
<summary>Old layout:</summary>

Simulator Screenshot - iPhone 15 Pro - 2023-11-17 at 16 19 34


</details>

<details>
<summary>Layout with these suggestions:</summary>

Simulator Screenshot - iPhone 15 Pro - 2023-11-17 at 16 18 39


</details>

Comment (s77rt) [reply src/components/Form/FormProvider.js:77]: True but I think the new layout is consist with the rest of the Forms in the App. The error placement can be moved to be below the label.

Screenshot 2023-11-17 at 4 25 57 PM

Comment (s77rt) [reply src/components/Form/FormProvider.js:77]: cc @shawnborton We are going to make a slight change in the error position (see old layout vs new layout screenshots above) for consistency.

Should the error be below the label as in https://user-images.githubusercontent.com/16493223/283849311-247c0f78-7f8f-4347-9a5b-a8fa054f2d89.png or below the questions as in https://user-images.githubusercontent.com/91068263/283847564-cbf83ccb-36e0-4abb-b622-88429c7dae8e.png

File: src/components/Form/FormProvider.js
Comment (kowczarz) [reply src/components/Form/FormProvider.js]: Yes, the function runs each time the input is re-rendered, so it's likely going to happen that it will be called after the user inputs something.

Comment (s77rt) [reply src/components/Form/FormProvider.js:219]: I see but what about `shouldUseDefaultValue`. We don't use that condition there

File: src/components/Form/FormProvider.js
Comment (kowczarz) [reply src/components/Form/FormProvider.js:77]: @shawnborton @luacmartins do you have any thoughts on this?

Comment (shawnborton) [reply src/components/Form/FormProvider.js:77]: I think below the questions makes more sense to me.

Comment (shawnborton) [reply src/components/Form/FormProvider.js:77]: cc @Expensify/design in case they have a different opinion though!

Comment (dannymcclain) [reply src/components/Form/FormProvider.js:77]: I don't have a strong opinion—below the question or the options works for me. I think either is a good improvement over the old layout!

Comment (kowczarz) [reply src/components/Form/FormProvider.js:77]: So are we good to go with the error below the questions?

Comment (luacmartins) [reply src/components/Form/FormProvider.js:77]: @kowczarz it seems like we have design approval!

Comment (kowczarz) [reply src/components/Form/FormProvider.js:77]: Here you can see how it looks now on web/desktop:
Screenshot 2023-11-22 at 16 48 03

and on phones:
Simulator Screenshot - iPhone 15 Pro - 2023-11-22 at 16 48 22

File: src/components/Form/FormProvider.js
Comment (kowczarz) [reply src/components/Form/FormProvider.js:52]: @cdOut had to implement same thing in https://github.com/Expensify/App/pull/30525, so probably I will wait until it will be merged.

File: src/pages/EnablePayments/IdologyQuestions.js
Comment (s77rt) [inline src/pages/EnablePayments/IdologyQuestions.js:145]: Remove this View

File: src/components/SingleChoiceQuestion.tsx
Comment (s77rt) [inline src/components/SingleChoiceQuestion.tsx:42]:
```suggestion
SingleChoiceQuestion.displayName = 'SingleChoiceQuestion';

export default forwardRef(SingleChoiceQuestion);
```

File: src/components/SingleChoiceQuestion.tsx
Comment (s77rt) [inline src/components/SingleChoiceQuestion.tsx:22]: How is this causing a crash? The FormWrapper checks if `focus` method exists before calling it

**Review by s77rt — CHANGES_REQUESTED** [review]

**Comment by s77rt** [comment]
@kowczarz Please fix the passed form id in `FormActions.setDraftValues`. It should be `formID` not `propsToParse.formID`.  `propsToParse` are the input props [^1].

[^1]: https://github.com/Expensify/App/issues/31612#issuecomment-1820779926

**Comment by dubielzyk-expensify** [comment]
Let's make sure the text and checkboxes are aligned with everything else:
CleanShot 2023-11-23 at 06 46 41@2x

File: src/components/SingleChoiceQuestion.tsx
Comment (kowczarz) [reply src/components/SingleChoiceQuestion.tsx]: I thought so, but on iPhone this appeared when I call `ref.current.focus()`. We are checking if it exists in JS, but it seems the native implementation doesn't support focusing on `Text`.
Simulator Screenshot - iPhone 15 Pro - 2023-11-23 at 15 02 42

Comment (kowczarz) [reply src/components/SingleChoiceQuestion.tsx]: Ok, then I'm removing it.

Comment (s77rt) [reply src/components/SingleChoiceQuestion.tsx]: This seems to be a bug on react native. Let's revert that change

**Comment by s77rt** [comment]
@kowczarz Please check https://github.com/Expensify/App/pull/31401#issuecomment-1823276310

**Comment by kowczarz** [comment]
@s77rt it's already fixed on `main` and I have already pulled the changes. I'm checking https://github.com/Expensify/App/issues/31612 and if it will work, we will be ready to go.

**Comment by s77rt** [comment]
Oh my bad. Code looks good to me then. Just gotta complete the checklist/tests

**Comment by s77rt** [comment]
All working good except the date of birth is not being saved. This is because the `NewDatePicker` is not being registered correctly as a from input. Currently the `NewDatePicker` internally uses `InputWrapper`. This is bad because any changes to the `NewDatePicker` itself won't trigger any form related action as we only registered the `TextInput` (via `InputWrapper`). This makes the `onInputChange` function on `NewDatePicker` useless (`onInputChange` is used to save the draft values).
 
**Current structure:**
```js
// IdentifyForm:
<NewDatePicker ... />

// NewDatePicker:
<InputWrapper InputComponent={TextInput} ... />
```


**Suggested structure:**
```js
// IdentifyForm:
<InputWrapper InputComponent={NewDatePicker} ... />

// NewDatePicker:
<TextInput ... /> 
```

I think this should be fixed in a follow up PR. Related PR: https://github.com/Expensify/App/pull/29098.

**Comment by s77rt** [comment]
## Reviewer Checklist

- [X] I have verified the author checklist is complete (all boxes are checked off).
- [X] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [X] I verified the steps for local testing are in the `Tests` section
    - [X] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [X] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [X] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [X] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [X] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [X] I verified tests pass on **all platforms** & I tested again on:
    - [X] Android: Native
    - [X] Android: mWeb Chrome
    - [X] iOS: Native
    - [X] iOS: mWeb Safari
    - [X] MacOS: Chrome / Safari
    - [X] MacOS: Desktop
- [X] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [X] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [X] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [X] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [X] I verified that comments were added to code that is not self explanatory
    - [X] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [X] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [X] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [X] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [X] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [X] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [X] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [X] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [X] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [X] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [X] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [X] If a new component is created I verified that:
    - [X] A similar component doesn't exist in the codebase
    - [X] All props are defined accurately and each prop has a `/** comment above it */`
    - [X] The file is named correctly
    - [X] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [X] The only data being stored in the state is data necessary for rendering and nothing else
    - [X] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [X] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [X] All JSX used for rendering exists in the render method
    - [X] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [X] If any new file was added I verified that:
    - [X] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [X] If a new CSS style is added I verified that:
    - [X] A similar style doesn't already exist
    - [X] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [X] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [X] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [X] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [X] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [X] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [X] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>


android

</details>

<details>
<summary>Android: mWeb Chrome</summary>


mweb-chrome

</details>

<details>
<summary>iOS: Native</summary>


ios

</details>

<details>
<summary>iOS: mWeb Safari</summary>


mweb-safari

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>


web

</details>

<details>
<summary>MacOS: Desktop</summary>


desktop

</details>

**Comment by s77rt** [comment]
This is only missing the testing steps and should be good to merge

**Comment by kowczarz** [comment]
@s77rt I think it's ready

**Comment by s77rt** [comment]
@kowczarz Thanks!

**Review by s77rt — APPROVED** [review]

**Comment by luacmartins** [comment]
@kowczarz @s77rt reassure test is failing

**Comment by luacmartins** [comment]
I'm gonna merge this with failing reassure test since the tests actually seem to pass, but there seems to be an unrelated error somewhere in the test script:

Screenshot 2023-11-27 at 10 56 08 AM

**Review by luacmartins — APPROVED** [review]

**Comment by luacmartins** [comment]
See above ^

**Comment by situchan** [comment]
Seems like this PR caused several deploy blockers ⬆️

**Comment by luacmartins** [comment]
@kowczarz @s77rt we reverted this PR because it introduced the blockers below. @kowczarz could you please work on a v2 PR taking the blockers into consideration? Thanks!

- https://github.com/Expensify/App/issues/32083
- https://github.com/Expensify/App/issues/32104

**Pull request merged**
